File: Programmers/KAKAO BLIND RECRUITMENT/2018/[1차] 뉴스 클러스터링/solution.py
# https://school.programmers.co.kr/learn/courses/30/lessons/17677


from collections import Counter
import re


def solution(str1, str2):
    l1, l2 = [], []
    # 정규식으로 알파벳만 골라냄
    for idx in range(len(str1) - 1):
        if re.match('[a-z]{2}', str1[idx:idx + 2].lower()):
            l1.append(str1[idx:idx + 2].lower())

    for idx in range(len(str2) - 1):
        if re.match('[a-z]{2}', str2[idx:idx + 2].lower()):
            l2.append(str2[idx:idx + 2].lower())

    c1, c2 = Counter(l1), Counter(l2)

    # and와 or 연산자로 합집합과 교집합을 구함
    intersection = sum((c1 & c2).values())
    union = sum((c1 | c2).values())

    jaccard = 1 if union == 0 else intersection / union

    return int(jaccard * 65536)

File: Programmers/KAKAO BLIND RECRUITMENT/2018/[1차] 뉴스 클러스터링/solution.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/17677?language=cpp

#include <string>
#include <regex>
#include <vector>
#include <iostream>
#include <algorithm>
#include <set>

using namespace std;

void split(multiset<string> &titleSet, string &title)
{
    regex word("[a-z]{2}"); // 정규식으로 영단어 2글자만 고름
    string subTitle;
    for (int idx = 0; idx < title.size() - 1; idx++)
    {
        subTitle = title.substr(idx, 2);
        transform(subTitle.begin(), subTitle.end(), subTitle.begin(), ::tolower); // 대소문자 구분을 안하므로 다 소문자로 만듬
        if (regex_match(subTitle, word))                                          // 자른 2글자가 영어로만 이루어져있을경우에만 추가
        {
            titleSet.insert(subTitle);
        }
    }
}

int solution(string str1, string str2)
{
    multiset<string> setStr1, setStr2;
    split(setStr1, str1);
    split(setStr2, str2);
    // 교집합 합집합만들기
    multiset<string> intersection, setUnion;
    set_intersection(setStr1.begin(), setStr1.end(), setStr2.begin(), setStr2.end(), inserter(intersection, intersection.begin()));
    set_union(setStr1.begin(), setStr1.end(), setStr2.begin(), setStr2.end(), inserter(setUnion, setUnion.begin()));
    // 테스트 코드
    // for (auto s : intersection)
    // {
    //     cout << s << endl;
    // }
    // cout << "-----" << endl;
    // for (auto s : setUnion)
    // {
    //     cout << s << endl;
    // }

    // Divide by zero방지
    return setUnion.size() == 0 ? 65536 : (float)intersection.size() / setUnion.size() * 65536;
}